overlord/snapstate/backend/setup.go

```diff
@@ -111,6 +114,67 @@ func (b Backend) SetupSnap(snapFilePath, instanceName string, sideInfo *snap.Sid
 	return t, installRecord, nil
 }
 
+func earlyKernelMountDir(ksnapName string, rev snap.Revision) string {
+	return filepath.Join("/run/mnt/kernel-snap", ksnapName, rev.String())
+}
+
+// SetupKernelSnap does extra configuration for kernel snaps.
+func (b Backend) SetupKernelSnap(instanceName string, rev snap.Revision, meter progress.Meter) (err error) {
+	var sysd systemd.Systemd
+	if b.preseed {
+		sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
+	} else {
+		sysd = systemd.New(systemd.SystemMode, meter)
+	}
+
+	// Create early mount for the snap
+	cpi := snap.MinimalSnapContainerPlaceInfo(instanceName, rev)
+
+	earlyMountDir := earlyKernelMountDir(instanceName, rev)
+	addMountUnitOptions := &systemd.MountUnitOptions{
+		MountUnitType: systemd.BeforeDriversLoadMountUnit,
+		Lifetime:      systemd.Persistent,
+		Description:   "Early mount unit for kernel snap",
+		What:          cpi.MountFile(),
+		Where:         earlyMountDir,
+		Fstype:        "squashfs",
+		Options:       []string{"nodev,ro,x-gdu.hide,x-gvfs-hide"},
+	}
+	_, err = sysd.EnsureMountUnitFileWithOptions(addMountUnitOptions)
```

Comment:
will this unit clash with what we are doing right now in snap-bootstrap?

Reply:
No, as this mounts the kernel in a different place (/run/mnt/kernel-snap/...). Where there is some conflict is in `/lib/{modules,firmware}` mounts, that are done by the initramfs, and we want to do instead from a systemd generator. But even in that case it won't be a problem in case an old initramfs is used with a new base as the old mounts would be clobbered by the new ones.

Reply:
thanks, I thought so, what wanted to double check